README.md
Check_MK plugin for Citrix SD-WAN
=================================


About
-----

This Check_MK package monitors the WAN status of Citrix SD-WAN appliances.


ChangeLog
cmk_ctx-sdwan (0.1) unstable; urgency=low

  * Initial release.

 -- Thomas Liske <liske@ibh.de>


checks/ctx_sdwan_ha.py
#!/usr/bin/env python3

# cmk_ctx-sdwan - check-mk plugin for SNMP-based Cisco Digital-Optical-Monitoring monitoring
#
# Authors:
#   Thomas Liske <liske@ibh.de>
#	Philipp Kilian <kilian@ibh.de>
#
# Copyright Holder:
#   2022 (C) IBH IT-Service GmbH [http://www.ibh.de/]
#
# License:
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this package; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#

from typing import List, Any, Mapping
from .agent_based_api.v1 import register, Result, Service, SNMPTree, State, Metric, startswith
from .agent_based_api.v1.type_defs import CheckResult, StringTable, InventoryResult


ctx_sdwan_ha_states = {
    0: 'undefined',
    1: 'notConfigured',
    2: 'active',
    3: 'standby'
}


def parse_ctxsdwan_ha(string_table: List[StringTable]) -> int:
    if string_table[0][0][0]:
        return int(string_table[0][0][0])
    else:
        return 0


register.snmp_section(
    name="ctx_sdwan_ha",
    parse_function=parse_ctxsdwan_ha,
    fetch=[
        SNMPTree(
            base='.1.3.6.1.4.1.3845.31.4.2.2.12.1',
            oids=['9']
        )
    ],
    detect=startswith(".1.3.6.1.2.1.1.2.0", ".1.3.6.1.4.1.3845.31")
)


def discovery_ctxsdwan_ha(section: int) -> InventoryResult:
    if section > 1:
        yield Service(item="HA State", parameters={"discovery_value": section})


def check_ctxsdwan_ha(item, params: Mapping[str, Any], section: int) -> CheckResult:
    c_state = State.OK
    if section != params["discovery_value"]:
        c_state = State.CRIT
        if section > 1:
            c_state = State.WARN

    yield Result(state=c_state, summary=f"HA state is {ctx_sdwan_ha_states[section]} ({section})")
    yield Metric('ha_state', section)


register.check_plugin(
    name="ctx_sdwan_ha",
    service_name="SD-WAN %s",
    discovery_function=discovery_ctxsdwan_ha,
    check_function=check_ctxsdwan_ha,
    check_default_parameters={"discovery_value": 0}
)


checks/ctx_sdwan_link.py
#!/usr/bin/env python3

# cmk_ctx-sdwan - check-mk plugin for SNMP-based Cisco Digital-Optical-Monitoring monitoring
#
# Authors:
#   Thomas Liske <liske@ibh.de>
#	Philipp Kilian <kilian@ibh.de>
#
# Copyright Holder:
#   2022 (C) IBH IT-Service GmbH [http://www.ibh.de/]
#
# License:
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this package; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#

import time
from typing import List, Any, Mapping, NamedTuple
from .agent_based_api.v1 import register, Result, Service, SNMPTree, State, Metric, startswith, get_rate, OIDEnd, get_value_store
from .agent_based_api.v1.type_defs import CheckResult, StringTable, InventoryResult

ctx_sdwan_link_states = {
    0: [State.OK, 'undefined'],
    1: [State.WARN, 'disabled'],
    2: [State.CRIT, 'dead'],
    3: [State.WARN, 'bad'],
    4: [State.OK, 'good'],
}


class SDWANLink(NamedTuple):
    name: str
    idx: int
    state: int
    traffic_data: dict


def parse_ctxsdwan_link(string_table: List[StringTable]) -> List[SDWANLink]:
    links = []
    for link in string_table[0]:
        links.append(
            SDWANLink(
                name=link[1], 
                idx=int(link[0]), 
                state=int(link[2]),
                traffic_data={
                    'bps_tx': int(link[3]),
                    'pps_tx': int(link[4]),
                    'bps_rx': int(link[5]),
                    'pps_rx': int(link[6]),
                    'bps_drop': int(link[7]),
                    'pps_drop': int(link[8])
                }
            )
        )
    return links


register.snmp_section(
    name="ctx_sdwan_link",
    parse_function=parse_ctxsdwan_link,
    fetch=[
        # CITRIX-SD-WAN-MIB::sdWANStatsWANLinkEntry
        SNMPTree(
            base='.1.3.6.1.4.1.3845.31.4.2.2.15.2.1',
            oids=[OIDEnd(), '3', '4', '5', '6', '7', '8', '9', '10']
        )
    ],
    detect=startswith(".1.3.6.1.2.1.1.2.0", ".1.3.6.1.4.1.3845.31")
)


def discovery_ctxsdwan_link(section: List[SDWANLink]) -> InventoryResult:
    for link in section:
        yield Service(item=link.name, parameters={"discovery_idx": link.idx})


# eval service state
def check_ctxsdwan_link(item, params: Mapping[str, int], section: List[SDWANLink]) -> CheckResult:
    for link in section:
        if link.name == item and link.idx == params["discovery_idx"]:
            if link.state not in ctx_sdwan_link_states:
                yield Result(state=State.UNKNOWN, summary=f"Link state '{link.state}' is unknown")
                break

            yield Result(state=ctx_sdwan_link_states[link.state][0], summary=f"Link state is {ctx_sdwan_link_states[link.state][1]} ({link.state})")
            this_time = int(time.time())
            yield Metric("state", link.state)
            value_store = get_value_store()
            for name, value in link.traffic_data.items():
                yield Metric(name, get_rate(value_store, f"SD-WAN-Link.{link.idx}.{name}", this_time, value))
            break


register.check_plugin(
    name="ctx_sdwan_link",
    service_name="SD-WAN Link %s",
    discovery_function=discovery_ctxsdwan_link,
    check_function=check_ctxsdwan_link,
    check_default_parameters={"discovery_idx": 0}
)


checks/ctx_sdwan_vpath.py
#!/usr/bin/env python3

# cmk_ctx-sdwan - check-mk plugin for SNMP-based Cisco Digital-Optical-Monitoring monitoring
#
# Authors:
#   Thomas Liske <liske@ibh.de>
#	Philipp Kilian <kilian@ibh.de>
#
# Copyright Holder:
#   2022 (C) IBH IT-Service GmbH [http://www.ibh.de/]
#
# License:
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this package; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#

import time
from typing import List, Mapping, NamedTuple
from .agent_based_api.v1 import register, Result, Service, SNMPTree, State, Metric, startswith, get_rate, get_value_store, OIDEnd
from .agent_based_api.v1.type_defs import CheckResult, StringTable, InventoryResult


ctx_sdwan_vpath_states = {
    0: [State.OK, 'undefined'],
    1: [State.CRIT, 'disabled'],
    2: [State.CRIT, 'dead'],
    3: [State.CRIT, 'bad'],
    4: [State.OK, 'good'],
}


class SDWANVPath(NamedTuple):
    name: str
    idx: int
    state: int
    path_data: dict


def parse_ctxsdwan_path(string_table: List[StringTable]) -> List[SDWANVPath]:
    paths = []
    for path in string_table[0]:
        paths.append(
            SDWANVPath(
                idx=int(path[0]),
                name=path[1],
                state=int(path[2]),
                path_data={
                    'tx_bps': [float(path[3]), 1, True],
                    'tx_pps': [float(path[4]), 1, True],
                    'rx_bps': [float(path[5]), 1, True],
                    'rx_pps': [float(path[6]), 1, True],
                    'paths': [float(path[7]), 1, False],
                    'rules': [float(path[8]), 1, False],
                    'tx_bps_drop': [float(path[9]), 1, True],
                    'tx_pps_drop': [float(path[10]), 1, True],
                    'tx_pps_lost': [float(path[11]), 1, True],
                    'tx_pps_ooo': [float(path[12]), 1, True],
                    'tx_bowt': [float(path[13]), .1, False],
                    'tx_jitter': [float(path[14]), .1, False],
                    'rx_bps_drop': [float(path[15]), 1, True],
                    'rx_pps_drop': [float(path[16]), 1, True],
                    'rx_pps_lost': [float(path[17]), 1, True],
                    'rx_pps_ooo': [float(path[18]), 1, True],
                    'rx_bowt': [float(path[19]), .1, False],
                    'rx_jitter': [float(path[20]), .1, False]
                }
            )
        )
    return paths


register.snmp_section(
    name="ctx_sdwan_vpath",
    parse_function=parse_ctxsdwan_path,
    fetch=[
        SNMPTree(
            # CITRIX-SD-WAN-MIB::sdWANStatsVPathEntry
            base='.1.3.6.1.4.1.3845.31.4.2.2.16.2.1',
            oids=[OIDEnd(), '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22']
        )
    ],
    detect=startswith(".1.3.6.1.2.1.1.2.0", ".1.3.6.1.4.1.3845.31")
)


# generate list of valid monitors
def discovery_ctxsdwan_vpath(section: List[SDWANVPath]) -> InventoryResult:
    for path in section:
        yield Service(item=path.name, parameters={"discovery_idx": path.idx})


# eval service state
def check_ctxsdwan_vpath(item, params: Mapping[str, int], section: List[SDWANVPath]) -> CheckResult:
    for path in section:
        if item == path.name and path.idx == params["discovery_idx"]:
            if path.state not in ctx_sdwan_vpath_states:
                yield Result(state=State.UNKNOWN, summary=f"Path state '{path.state}' is unknown")
                break

            yield Result(state=ctx_sdwan_vpath_states[path.state][0], summary=f"VPath state is {ctx_sdwan_vpath_states[path.state][1]}({path.state})")
            this_time = int(time.time())
            yield Metric('state', path.state)
            value_store = get_value_store()

            for name, opts in path.path_data.items():
                if opts[2]:
                    replaced_name = name.replace('->', '_').replace('<-', '_')
                    yield Metric(name, get_rate(value_store, f"SD-WAN-VPath.{path.idx}.{replaced_name}", this_time, opts[0] * opts[1]))
                else:
                    yield Metric(name, opts[0] * opts[1])
            break


register.check_plugin(
    name="ctx_sdwan_vpath",
    service_name="SD-WAN VPath %s",
    discovery_function=discovery_ctxsdwan_vpath,
    check_function=check_ctxsdwan_vpath,
    check_default_parameters={"discovery_idx": 0}
)


checks/ctx_sdwan_path.py
#!/usr/bin/env python3

# cmk_ctx-sdwan - check-mk plugin for SNMP-based Cisco Digital-Optical-Monitoring monitoring
#
# Authors:
#   Thomas Liske <liske@ibh.de>
#	Philipp Kilian <kilian@ibh.de>
#
# Copyright Holder:
#   2022 (C) IBH IT-Service GmbH [http://www.ibh.de/]
#
# License:
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this package; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
#

import time
from typing import List, Mapping, NamedTuple
from .agent_based_api.v1 import register, Result, Service, SNMPTree, State, Metric, startswith, get_rate, get_value_store
from .agent_based_api.v1.type_defs import CheckResult, StringTable, InventoryResult

ctx_sdwan_path_states = {
    0: [State.OK, 'undefined'],
    1: [State.CRIT, 'disabled'],
    2: [State.CRIT, 'dead'],
    3: [State.CRIT, 'bad'],
    4: [State.OK, 'good'],
}


class SDWANPath(NamedTuple):
    name: str
    idx: int
    state: int
    path_data: dict


def parse_ctxsdwan_path(string_table: List[StringTable]) -> List[SDWANPath]:
    paths = []
    for path in string_table[0]:
        paths.append(
            SDWANPath(
                idx=int(path[0]),
                name=path[1],
                state=int(path[2]),
                path_data={
                    'tx_bps': [float(path[3]), 1, True],
                    'tx_pps': [float(path[4]), 1, True],
                    'rx_bps': [float(path[5]), 1, True],
                    'rx_pps': [float(path[6]), 1, True],
                    'bowt': [float(path[7]), 0.1, False],
                    'jitter': [float(path[8]), 0.1, False],
                    'pps_lost': [float(path[9]), 1, True],
                    'pps_ooo': [float(path[10]), 1, True]
                }
            )
        )
    return paths


register.snmp_section(
    name="ctx_sdwan_path",
    parse_function=parse_ctxsdwan_path,
    fetch=[
        SNMPTree(
            # CITRIX-SD-WAN-MIB::sdWANStatsWANPathEntry
            base='.1.3.6.1.4.1.3845.31.4.2.2.16.3.1.1',
            oids=['2', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14']
        )
    ],
    detect=startswith(".1.3.6.1.2.1.1.2.0", ".1.3.6.1.4.1.3845.31")
)


def discovery_ctxsdwan_path(section: List[SDWANPath]) -> InventoryResult:
    for path in section:
        yield Service(item=path.name, parameters={"discovery_idx": path.idx})


# eval service state
def check_ctxsdwan_path(item, params: Mapping[str, int], section: List[SDWANPath]) -> CheckResult:
    for path in section:
        if item == path.name and path.idx == params["discovery_idx"]:
            if path.state not in ctx_sdwan_path_states:
                yield Result(state=State.UNKNOWN, summary=f"Path state '{path.state}' is unknown")
                break

            yield Result(state=ctx_sdwan_path_states[path.state][0], summary=f"Path state is {ctx_sdwan_path_states[path.state][1]}({path.state})")
            this_time = int(time.time())
            yield Metric('state', path.state)
            value_store = get_value_store()

            for name, opts in path.path_data.items():
                if opts[2]:
                    replaced_name = name.replace('->', '_').replace('<-', '_')
                    yield Metric(name, get_rate(value_store, f"SD-WAN-Path.{path.idx}.{replaced_name}", this_time, opts[0] * opts[1]))
                else:
                    yield Metric(name, opts[0] * opts[1])
            break


register.check_plugin(
    name="ctx_sdwan_path",
    service_name="SD-WAN Path %s",
    discovery_function=discovery_ctxsdwan_path,
    check_function=check_ctxsdwan_path,
    check_default_parameters={"discovery_idx": 0}
)
